x/distribution/simulation/invariants.go

```diff
@@ -48,3 +52,81 @@ func ValAccumInvariants(k distr.Keeper, sk distr.StakeKeeper) simulation.Invaria
 		return nil
 	}
 }
+
+// DelAccumInvariants checks that each validator del accum == sum all delegators' accum
+func DelAccumInvariants(k distr.Keeper, sk distr.StakeKeeper) simulation.Invariant {
```

Comment:
This is a great invariant but it is *very* slow, I wonder if we should make it periodic by default.

Reply:
I think we should leave it at every block until we upgrade for more capable feature set in the simulator - From a debugging perspective it's nice to not have to be an expert and just "know" that we should reset an invariant to be running each block when it fails to determine the location of the bug. I'm also not so worried about the speed of the simulator - this doesn't seem to be an issue as I've been running things locally. 

To improve the speed let's focus on things like the ability to load the simulator from a simulated state etc. 

As per running this every block _on chain_ - yeah that seems like a performance issue - not sure how to adjust that (is this per block invariants stuff already merged? - feel free to push here to make that fix, or we could add another PR)

Reply:
I think we should run this on-chain and not run it for blockchain syncing for new nodes... I think it would be nice to halt the chain as soon as we know that something is off, at least in the beginning.

Reply:
:100: - we can do this very easily, just edit the runtime invariants in `cmd/gaia/app/invariants.go`

Reply:
(right now we aren't running it on-chain @rigelrozanski, you have to add it explicitly)

Reply:
>  I'm also not so worried about the speed of the simulator - this doesn't seem to be an issue as I've been running things locally.

OK, maybe it's just my slow laptop - but we should ensure we can still run 500/1000-block simulations as they do catch issues.

Reply:
Well, depends on how fast it is... checking...

Reply:
ahhhhhhs didn't realize _how_ much slower this is

I think that we should probably add an option to just "enable" or "disable" slow invar checks for the simulation here... maybe a flag

Reply:
Hmm I much prefer periodic invariants, which achieve essentially the same speed benefit while still ensuring errors are caught (and once caught, you can disable periodicity to figure out the exact operation).